### PR TITLE
Optimize peeks

### PR DESCRIPTION
**Pull request by xwipeoutx**
# Description

Fixes:

- #1614

This reduces the number of times the transport will run the "receive" query unnecessarily by ensuring that messages in the current "peek" batch have all at least locked the message row (via the "receive") query before doing another peek operation.

In the case of a single physical endpoint, this completely eliminates redundant reads.

In the case of multiple physical endpoints, there will still be redundant reads due to competing consumers, but this is still a significant improvement.

## Results

In testing on real-world systems, we have 300% unnecessary additional "receive" calls when the queue was empty, or all other messages were being processed. This PR makes that 0.

On a synthetic test we did, with multiple physical endpoints, each endpoint saw a similar reduction in this overrun.

## Approach

A new concept `AsyncCountdownLatch` has been added, which is similar to a `CountdownEvent` but works with async.

The message receiver creates one of these for each peek batch, with a count equal to the batch size, and passes it down to the `ProcessStrategy` being used.

The strategy will signal this latch when its receive query has been completed, and the latch will be released when all have been completed.

There is an edge case accounted for when the processing is cancelled - in this case, we want to ignore the latch, and exit the method anyway.

## Quality

All existing integration and unit tests still pass.  None extra have been written, I'm open to guidance from Particular on this.

### CODE REVIEW

File: src/NServiceBus.Transport.Sql.Shared/Receiving/MessageReceiver.cs [hunk 1]
```diff
@@ -216,13 +219,19 @@ async Task ReceiveMessages(CancellationToken messageReceivingCancellationToken)
...
 
                 _ = ProcessMessagesSwallowExceptionsAndReleaseConcurrencyLimiter(stopBatchCancellationSource,
-                    localConcurrencyLimiter, messageProcessingCancellationTokenSource.Token);
+                    localConcurrencyLimiter, receiveLatch, messageProcessingCancellationTokenSource.Token);
+            }
+
+            if (shouldWaitForReceiveTasks)
+            {
+                // Wait for all receive operations to complete before returning (and thus peeking again)
+                await receiveLatch.WaitAsync().ConfigureAwait(false);
```
Comment (tmasternak) [inline src/NServiceBus.Transport.Sql.Shared/Receiving/MessageReceiver.cs]: We want to pass `messageReceivingCancellationToken` here, I think.

**Review by tmasternak — COMMENTED** [review]
Overall, this looks good to me!

One comment regarding:
> There is an edge case accounted for when the processing is cancelled - in this case, we want to ignore the latch, and exit the method anyway.

The `stopBatchCancellationSource` is used to detect a scenario when a peek saw more messages than are actually available in the queue (due to some other process reading the messages already). In such a case, we don't want to continue scheduling `receive` operations but re-execute the peek operation instead to get a better estimate on the input queue size.

Secondly, could you share which database engine (MS SQL or PostgreSQL) you are running your system on? 

Thank you :).

**Comment by mauroservienti** [comment]
@xwipeoutx 

> @mauroservienti Do you want me to continue to contribute here? Or will you take the PR from here?

It'd be better if you contribute here; I'll take care of the final steps to get this merged once it's approved. We cannot run CI on this PR because it won't have access to the required secrets. So, once this is ready, I'll cherry-pick all your commits into a new PR to keep your name on them, while enabling CI to run since it'll be me triggering the build.

**Comment by xwipeoutx** [comment]
Brilliant.

I'm seeing some discussion here which I can't contribute strongly too - but it seems the overall shape is fine? Are there any particulars preventing this from being approved in its current state?

File: src/NServiceBus.Transport.Sql.Shared/Receiving/ProcessWithNativeTransaction.cs [hunk 2]
```diff
@@ -31,7 +32,16 @@ public override async Task ProcessMessage(CancellationTokenSource stopBatchCance
...
-                    var receiveResult = await InputQueue.TryReceive(connection, transaction, cancellationToken).ConfigureAwait(false);
+                    MessageReadResult receiveResult;
+                    try
+                    {
+                        receiveResult = await InputQueue.TryReceive(connection, transaction, cancellationToken)
+                            .ConfigureAwait(false);
+                    }
+                    finally
+                    {
+                        receiveLatch.Signal();
```
Comment (SzymonPobiega) [inline src/NServiceBus.Transport.Sql.Shared/Receiving/ProcessWithNativeTransaction.cs]: This probably needs to be somehow moved up because in the current shape if an exception happens when opening connection or transaction and all other concurrent receives succeed, the counter will be missing one element and won't let the peek continue.
